### PR TITLE
Java: add models of JSON-java, aka `org.json`

### PR DESCRIPTION
**Pull request by smowton**
Strategy: like `javax.json`, I have adopted the approach of associated monolithic taint with JSON objects as a whole rather than elements, map-values and so on, on the assumption that 99% of use cases are about de/serialization and not using the objects like containers.

### CODE REVIEW

File: java/ql/test/library-tests/frameworks/json-java/output.bqrs [hunk 1]
```diff
@@ -0,0 +1,528 @@
+��org.json.CDL�// Generated automatically from org.json.CDL for testing purposes
```
Comment (aschackmull) [inline java/ql/test/library-tests/frameworks/json-java/output.bqrs:1]: Committed by accident?

File: java/ql/test/library-tests/frameworks/json-java/output.json [hunk 2]
```diff
@@ -0,0 +1,24 @@
+{"#select":{"columns":[
```
Comment (aschackmull) [inline java/ql/test/library-tests/frameworks/json-java/output.json:1]: Committed by accident?

File: java/ql/test/library-tests/frameworks/json-java/output.bqrs (hunk 1 above)
Comment (smowton) [reply java/ql/test/library-tests/frameworks/json-java/output.bqrs]: Doh, yes, removed.

**Review by Marcono1234 — COMMENTED** [review]
Great that this pull request covers the complete JSON-java project!

Hopefully the following review comments are helpful.

File: java/ql/src/semmle/code/java/frameworks/JsonJava.qll [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+        "org.json;JSONPointer;false;JSONPointer;(List);;Element of Argument[0];Argument[-1];taint",
+        "org.json;JSONPointer;false;JSONPointer;(String);;Argument[0];Argument[-1];taint",
+        "org.json;JSONPointer;false;queryFrom;;;Argument[0];ReturnValue;taint",
+        "org.json;JSONPointer;false;toString;;;Argument[-1];ReturnValue;taint",
+        "org.json;JSONPointer;false;toURIFragment;;;Argument[-1];ReturnValue;taint",
+        "org.json;JSONPointer$Builder;false;append;;;Argument[0];Argument[-1];taint",
+        "org.json;JSONPointer$Builder;false;append;;;Argument[-1];ReturnValue;value",
+        "org.json;JSONPointer$Builder;false;build;;;Argument[-1];ReturnValue;taint",
+        "org.json;JSONStringer;false;toString;;;Argument[-1];ReturnValue;taint",
+        "org.json;JSONTokener;false;JSONTokener;;;Argument[0];Argument[-1];taint",
```
[the remark this answers is not in the document]
Comment (smowton) [reply java/ql/src/semmle/code/java/frameworks/JsonJava.qll:205]: I left individual chars out as too far-fetched a taint vector

File: java/ql/src/semmle/code/java/frameworks/JsonJava.qll [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+        "org.json;JSONTokener;false;next;;;Argument[-1];ReturnValue;taint",
+        "org.json;JSONTokener;false;nextClean;;;Argument[-1];ReturnValue;taint",
+        "org.json;JSONTokener;false;nextString;;;Argument[-1];ReturnValue;taint",
+        "org.json;JSONTokener;false;nextTo;;;Argument[-1];ReturnValue;taint",
+        "org.json;JSONTokener;false;nextValue;;;Argument[-1];ReturnValue;taint",
+        "org.json;JSONTokener;false;syntaxError;;;Argument[0..1];ReturnValue;taint",
+        "org.json;JSONTokener;false;toString;;;Argument[-1];ReturnValue;taint",
+        "org.json;JSONWriter;true;JSONWriter;;;Argument[-1];Argument[0];taint",
```
[the remark this answers is not in the document]
Comment (smowton) [reply java/ql/src/semmle/code/java/frameworks/JsonJava.qll:213]: Not yet (Go can do this, Java can't). Will add a comment that this doesn't work yet.

File: java/ql/src/semmle/code/java/frameworks/JsonJava.qll [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+        "org.json;JSONWriter;true;key;;;Argument[-1];ReturnValue;value",
+        "org.json;JSONWriter;true;object;;;Argument[-1];ReturnValue;value",
+        "org.json;JSONWriter;true;value;;;Argument[-1];ReturnValue;value",
+        "org.json;Property;false;toProperties;;;Argument[0];MapKey of ReturnValue;taint",
+        "org.json;Property;false;toProperties;;;Argument[0];MapValue of ReturnValue;taint",
+        "org.json;Property;false;toJSONObject;;;MapKey of Argument[0];ReturnValue;taint",
+        "org.json;Property;false;toJSONObject;;;MapValue of Argument[0];ReturnValue;taint",
+        "org.json;XML;false;escape;;;Argument[0];ReturnValue;taint",
+        "org.json;XML;false;stringToValue;;;Argument[0];ReturnValue;taint",
+        "org.json;XML;false;toJSONObject;;;Argument[0];ReturnValue;taint",
```
[the remark this answers is not in the document]
Comment (smowton) [reply java/ql/src/semmle/code/java/frameworks/JsonJava.qll:229]: I'll leave this as relevant taint flow seems very far-fetched

**Comment by smowton** [comment]
Thanks @Marcono1234 @aschackmull this is ready for re-review

File: java/ql/src/semmle/code/java/frameworks/JsonJava.qll (hunk 4 above)
[the remark this answers is not in the document]
Comment (aschackmull) [reply java/ql/src/semmle/code/java/frameworks/JsonJava.qll:225]: I'd be curious to know more about how this is designed to work in Go. As-is, this line is never going to work so might as well remove it (the input `Argument[-1]` to a constructor in Java is the result of the implicit malloc, which is passed into the constructor as the value of the `this` parameter, and there isn't really any way that flow can reach that `Node`).

File: java/ql/test/library-tests/frameworks/json-java/test.ql [hunk 6]
```diff
@@ -0,0 +1,53 @@
+import java
+import semmle.code.java.dataflow.DataFlow
+import semmle.code.java.dataflow.ExternalFlow
```
Comment (aschackmull) [inline java/ql/test/library-tests/frameworks/json-java/test.ql:3]:
```suggestion
```
No need for this import, right?

File: java/ql/src/semmle/code/java/frameworks/JsonJava.qll (hunk 4 above)
[the remark this answers is not in the document]
Comment (smowton) [reply java/ql/src/semmle/code/java/frameworks/JsonJava.qll:225]: The single-step relation (via the `FunctionInput` helper class) does some backward SSA use -> def walking, here: https://github.com/github/codeql-go/blob/main/ql/src/semmle/go/dataflow/FunctionInputsAndOutputs.qll#L124

For Java I expect this step would be something like: whenever taint should propagate from the result of a constructor (denoted `Argument[-1]`), introduce an edge from any post-update node concerning any variable that gets assigned the result of the constructor (or of the constructor itself if there is none) to the post-update node of the argument being tainted.

So here if we had

```
void f(Appendable sink, String source) {
  JSONWriter w = new JSONWriter(sink);
  w.write(source);
```

Then the `w.write(source)` would propagate taint from `source` to the post-update node of `w` as usual, but then the rule for `new JSONWriter(sink)` would wire that post-update node to `sink`'s post-update node (nb. Go's use of SSA definitions as the post-update nodes of corresponding uses means less SSA graph walking is necessary there).

This is pretty brittle in Go and would be here too, and is designed to pick up the simplest cases of writer-wrapping.

[the remark this answers is not in the document]
Comment (aschackmull) [reply java/ql/src/semmle/code/java/frameworks/JsonJava.qll:225]: I've put this in an issue for now: https://github.com/github/codeql/issues/6402

**Review by aschackmull — APPROVED** [review]

**Pull request merged**